libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -851,11 +865,27 @@ void IRGeneratorForStatements::endVisit(IndexAccess const& _indexAccess)
 				break;
 			}
 			case DataLocation::Memory:
-				solUnimplementedAssert(false, "");
-				break;
 			case DataLocation::CallData:
-				solUnimplementedAssert(false, "");
+			{
+				string const memAddress =
+					(arrayType.location() == DataLocation::Memory ?
+						m_utils.memoryArrayIndexAccessFunction(arrayType) :
+						m_utils.calldataArrayIndexAccessFunction(arrayType)) +
+					"(" +
+					m_context.variable(_indexAccess.baseExpression()) +
+					", " +
+					m_context.variable(*_indexAccess.indexExpression()) +
+					")";
+
+				setLValue(_indexAccess, make_unique<IRMemoryItem>(
```

Comment:
Actually calldata cannot be an lvalue...

Reply:
(because it is read-only)

Reply:
So calldata array index access will always be directly followed by `retrieveValue`, so maybe it makes sense to completely split it.

Reply:
I'll handle the calldata things in the next PR